Support for `NamespaceSelector`

zeeke opened the pull request:
This PR is about supporting `NamespaceSelector` for ingress/egress policies. I added a test with two pods in two different
namespace and I was expecting an iptables output similar to the pod selector one, but no rule gets generated.

The scenario is like the one described [here](https://github.com/ahmetb/kubernetes-network-policy-recipes/blob/master/06-allow-traffic-from-a-namespace.md).

If you confirm it's a bug, I can take the ownership of the fix with some suggestion

maiqueb commented:
I opened https://github.com/k8snetworkplumbingwg/multi-networkpolicy-iptables/issues/14 a few weeks ago, which raised (what I think is) a similar issue.

I could not validate it worked - it didn't work for my scenario - since I currently do not have enough time to investigate this issue, but this could never the less be helpful.

zeeke commented:
> I opened #14 a few weeks ago, which raised (what I think is) a similar issue.
> 
> I could not validate it worked - it didn't work for my scenario - since I currently do not have enough time to investigate this issue, but this could never the less be helpful.

Yes, it seems to be the same problem. Looking deeper, it seems iptables rules get generated if the namespace selector match the target pod namespace. So the problem is about having a  two namespaces (nsX, nsY), two pods (podA in nsX, podB in nsY) and a multinetwork-policy in nsX with podA as target and that reference nsY in ingress, egress rules.

I suppose it is almost the scenario of https://github.com/k8snetworkplumbingwg/multi-networkpolicy-iptables/issues/14

maiqueb commented:
> > I opened #14 a few weeks ago, which raised (what I think is) a similar issue.
> > I could not validate it worked - it didn't work for my scenario - since I currently do not have enough time to investigate this issue, but this could never the less be helpful.
> 
> Yes, it seems to be the same problem. Looking deeper, it seems iptables rules get generated if the namespace selector match the target pod namespace. So the problem is about having a two namespaces (nsX, nsY), two pods (podA in nsX, podB in nsY) and a multinetwork-policy in nsX with podA as target and that reference nsY in ingress, egress rules.

FWIW, my scenario did feature 2 different namespaces: I want *only* pods from namespace A to reach namespace B. 

> 
> I suppose it is almost the scenario of #14

@zeeke would you be willing to either re-open or create another issue, describing the detail about the namespace membership ?

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,92 @@ COMMIT
+		AddNamespace(s, "testns1")
+		AddNamespace(s, "testns2")
+
+		Expect(s.netdefChanges.Update(
+			nil,
+			NewNetDef("testns1", "net-attach1", NewCNIConfig("testCNI", "multi")))).To(BeTrue())
+		Expect(s.netdefChanges.GetPluginType(types.NamespacedName{Namespace: "testns1", Name: "net-attach1"})).To(Equal("multi"))
```
You also need to add net-attach-def, testns2/net-attach1 because pod2 uses testns2/net-attach1. Otherwise pod2 consumes non-exist net-attach-def and it causes inconsistency (it is not expected).

```
+               Expect(s.netdefChanges.Update(
+                       nil,
+                       NewNetDef("testns2", "net-attach1", NewCNIConfig("testCNI", "multi")))).To(BeTrue())
+               Expect(s.netdefChanges.GetPluginType(types.NamespacedName{Namespace: "testns2", Name: "net-attach1"})).To(Equal("multi"))
```

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,92 @@ COMMIT
+-A MULTI-0-INGRESS -j MULTI-0-INGRESS-0-FROM
+-A MULTI-0-INGRESS -m mark --mark 0x30000/0x30000 -j RETURN
+-A MULTI-0-INGRESS -j DROP
+-A MULTI-0-INGRESS-0-PORTS -i net1 -m tcp -p tcp --dport 8888 -j MARK --set-xmark 0x10000/0x10000
```
Your policy does not have 'ports' field, hence '-m tcp -p tcp --dport 8888' should not be generated. In this case, 
`-A MULTI-0-INGRESS-0-PORTS -m comment --comment "no ingress ports, skipped" -j MARK --set-xmark 0x10000/0x10000` should be generated.

zeeke replied:
You are right, this comes from a copy/paste from another testcase. going to fix it

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,92 @@ COMMIT
+
+		buf.renderIngress(s, podInfo1, 0, ingressPolicies1, []string{"testns1/net-attach1"})
+
+		//		fromRules := []byte("-A MULTI-0-INGRESS-0-FROM -i net1 -s 10.1.1.2 -j MARK --set-xmark 0x20000/0x20000\n")
+		//		Expect(buf.ingressFrom.String()).To(Equal(string(fromRules)))
+
```
I suppose it is for debug purpose. So please remove it if you finish to test.

zeeke replied:
Yes, of course

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,92 @@ COMMIT
+			nil)
+		AddPod(s, pod2)
+
+		buf.renderIngress(s, podInfo1, 0, ingressPolicies1, []string{"testns1/net-attach1"})
```
The last argument of renderIngress() specifies the target network (i.e. net-attach-def). In this case, pod2 attaches 'testns2/net-attach1', not 'testns1/net-attach1', hence you need to add 'testns2/net-attach1' as well, as following:

`buf.renderIngress(s, podInfo1, 0, ingressPolicies1, []string{"testns1/net-attach1", "testns2/net-attach1"})`

zeeke replied:
Got it, thank you for explaining.

zeeke commented:
I fixed the `ingress rules namespace selector` test case as suggested by @s1061123 : with two different net-attach-def we need to invoke `buf.renderIngress(...)` with both network attachment. Now it's green and I think it worths keeping it in the suite.

I also added another test case: `enforce policy with net-attach-def in a different namespace than pods` with the following scenario (@maiqueb  maybe it's more similar to your case):

- 3 namespaces: `default`, `testns1`, `testns2`
- 1 NetworkAttachmentDefinition `default/net-attach1`
- 2 pods with the above attachment: `testns1/testpod2` and `testns2/testpod2`
- 1 MultiNetworkPolicy with `testns/ingressPolicies1` with:
```
NamespaceSelector: &metav1.LabelSelector{
	MatchLabels: map[string]string{
		"nsname": "testns2",
	},
},
```
- `buf.renderIngress(s, podInfo1, 0, ingressPolicies1, []string{"default/net-attach1"})` as it had policy-for `default/net-attach1"`

It should be a consistent scenario, even though I'm not sure how frequent it can be in real use cases.

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,175 @@ COMMIT
+			"testns2",
+			"testpod2",
+			"default/net-attach1",
+			NewFakeNetworkStatus("testns2", "default/net-attach1", "192.168.1.2", "10.1.1.2"),
```
NewFakeNetworkStatus() takes argument 'net-attach-def namespace', 'net-attach-def name', 'eth0 ip', 'eth1 ip'.
Hence it should be `NewFakeNetworkStatus("default", "net-attach1", "192.168.1.2", "10.1.1.2"),` because you use net-attach-def in default namespace.

s1061123 commented on pkg/server/policyrules_test.go:
```diff
@@ -429,6 +431,175 @@ COMMIT
+:MULTI-0-INGRESS - [0:0]
+:MULTI-0-INGRESS-0-PORTS - [0:0]
+:MULTI-0-INGRESS-0-FROM - [0:0]
+-A MULTI-INGRESS -m comment --comment "policy:ingressPolicies1 net-attach-def:testns1/net-attach1" -i net1 -j MULTI-0-INGRESS
```
'net-attach-def:testns1/net-attach' should be 'net-attach-def:default/net-attach'

s1061123 commented:
@zeeke , thank you for your incorporate my comments. I suppose of course these tests are useful to improve CI, so let me merge that.
